discovery_engine_core/core/src/engine.rs

```diff
@@ -37,21 +37,42 @@ pub enum Error {
     /// Invalid stack: {0}.
     InvalidStack(#[source] stack::Error),
 
+    /// Invalid stack id: {0}.
+    InvalidStackId(StackId),
+
+    /// An operation on a stack failed: {0}.
+    StackOpFailed(#[source] stack::Error),
+
     /// Error while selecting the documents to return: {0}.
     Selection(#[from] mab::Error),
+
+    /// Error while using the ranker.
+    Ranker(#[from] GenericError),
+
+    /// A list of errors that could occur during some operation.
+    // This is often used when we want to run an operation on multiple data but
+    // we don't want to return on the first error.
+    // This is mainly done to maximize the amount of learning that we can do or
+    // try to run as many system to rank document as possible and discard only the one
+    // that are returning an error.
```

Comment:
personally I'd be in favour of keeping the comment - it's useful context I otherwise wouldn't have known.
```suggestion
    // try to run as many systems as possible to rank the documents, and discard only the ones
    // that are returning errors.
```